proposals/0000-hybrid-output.md

```diff
@@ -0,0 +1,139 @@
+
+The eventual solution should enable granular control over `output` on a per-route basis. Since Astro defaults to `'static'` output, this should remain the default in the eventual solution as well.
+
+# Detailed design
+
+## Public API
+
+There are two public API surfaces to this RFC:
+
+- Updating the `astro.config.mjs` to accept a new `output` value, `'hybrid'`. The existing values of `'static' | 'server'` would remain supported. A value of `'hybrid'` will enabled detection for the following feature:
+- Allowing routes (files in `pages/`) to self-declare a granular `output` value. The existing `output` values of `'static' | 'server'` would both be valid.
+
+**This declaration must be statically analyzable!**
+Only string literal values will be accepted. This is because a route cannot dynamically be static or server rendered based on an incoming request—this value _must_ be known at build time. Other declaration values will throw an error and fail the build. This mirrors similar constraints that Vite has for compile-time (macro) features like dynamic `import()` and `import.meta.glob`.
+
+```astro
+---
+// Valid – constant string literal
+export const output = 'server'
+// Invalid – cannot be determined at compile-time
+export const output = static ? 'static' : 'server';
+export const output = value;
+export const output = 'ser' + 'ver';
+export let output = undefined; output = 'server';
+---
+```
+
+## Internal API
+
+Astro needs to detect/understand the `export const output` syntax. This should be implemented as a specific post-processing Vite plugin. The plugin will detect the existence of valid ESM exports of `output` (`export const output`, `export let output`, `export var output`, `export { output }`, etc) and attach the statically detected `output` value to the module metadata under the `astro` namespace. This will allow us to track the `output` mode of any public route.
+
+> **Note**
+> Astro Routes include `.astro` pages, but also API endpoints (`src/pages/api/user.ts`)! It's important that this RFC handles both of these constraints in a uniform way, hence the use of a post-processing Vite plugin rather than implementing this feature in `@astrojs/compiler`.
+
+The **build** API needs to be updated to handle the new `'hybrid'` output mode. Implementation-wise, this will be similar to the existing `'server'` implementation, but it will also adopt portions of the `'static'` build to statically generate any routes that are not handled at request time.
+
+The **dev** server must be updated to allow `searchParams` and other features which are currently only avaliable with `server` rendering to be allowed on a per-route basis.
+
+## Adapter Support
+
+An adapter will be required when building for `hybrid` output, because we will be building a server with some statically generated routes. Adapters will need to be updated to handle `hybrid` output.
```

Comment:
Will adapters be expected to do anything differently for `static` routes, or the adapter only be envolved in the SSR routes?

Reply:
Open question, but my thinking is that adapters could safely ignore statically pre-rendered routes? They're essentially just `.html` files treated as assets.

Reply:
Updated the RFC to address this question. Adapters don't need to do anything differently, the generated assets will be included in the asset manifest and should be picked up automatically.